File: simulations-html/simulations/build-an-atom-html/src/app-easel/common/Vector2D.js
// Copyright 2002-2012, University of Colorado

/**
 * 2D vector.
 *
 * @author Chris Malley (PixelZoom, Inc.)
 */
define( [],
        function () {

            /**
             * @param {Number} x
             * @param {Number} y
             * @constructor
             */
            function Vector2D( x, y ) {
                this.x = x;
                this.y = y;
            }

            // @return {Number}
            Vector2D.prototype.getAngle = function () {
                return Math.atan2( this.y, this.x );
            };

            // @return {Number}
            Vector2D.prototype.getMagnitude = function () {
                return Math.sqrt( this.x * this.x + this.y * this.y );
            };

            // @return {String}
            Vector2D.prototype.toString = function () {
                return "[Vector2D (x=" + this.x + " y=" + this.y + ")]";
            };

            /**
             * Creates a vector using polar coordinates.
             * @param {Number} magnitude
             * @param {Number} angle in radians
             * @return {Vector$object}
             */
            Vector2D.createPolar = function ( magnitude, angle ) {
                var x = magnitude * Math.cos( angle );
                var y = magnitude * Math.sin( angle );
                return new Vector2D( x, y );
            };

            return Vector2D;
        } );

File: tags/faraday-v0r20-2005-03-29/faraday/src/edu/colorado/phet/faraday/util/Vector2D.java
/* Copyright 2005, University of Colorado */

/*
 * CVS Info -
 * Filename : $Source$
 * Branch : $Name$
 * Modified by : $Author$
 * Revision : $Revision$
 * Date modified : $Date$
 */

package edu.colorado.phet.faraday.util;

import java.awt.geom.Point2D;


/**
 * Vector2D is a mutable vector.
 *
 * @author Chris Malley (cmalley@pixelzoom.com)
 * @version $Revision$
 */
public class Vector2D {

    //----------------------------------------------------------------------------
    // Instance data
    //----------------------------------------------------------------------------
    
    private double _x;
    private double _y;
    private double _magnitude;
    private double _angle;

    //----------------------------------------------------------------------------
    // Constructors
    //----------------------------------------------------------------------------
    
    public Vector2D() {
        this( 0, 0 );
    }

    public Vector2D( double x, double y ) {
        setXY( x, y );
    }

    public Vector2D( Point2D p ) {
        this( p.getX(), p.getY() );
    }
    
    public Vector2D( Vector2D v ) {
        this( v.getX(), v.getY() );
    }

    public void copy( Vector2D v ) {
        _x = v.getX();
        _y = v.getY();
        _magnitude = v.getMagnitude();
        _angle = v.getAngle();
    }
    
    //----------------------------------------------------------------------------
    // XY accessors
    //----------------------------------------------------------------------------
    
    public void setXY( double x, double y ) {
        _x = x;
        _y = y;
        _magnitude = Math.sqrt( ( _x * _x ) + ( _y * _y ) );
        _angle = Math.atan2( _y, _x );
    }
    
    public void setX( double x ) {
        setXY( x, _y );
    }
    
    public void setY( double y ) {
        setXY( _x, y );
    }
    
    public double getX() {
        return _x;
    }
    
    public double getY() {
        return _y;
    }

    //----------------------------------------------------------------------------
    // Magnitude and Angle accessors
    //----------------------------------------------------------------------------
    
    public void setMagnitudeAngle( double magnitude, double angle ) {
        _magnitude = magnitude;
        _angle = angle;
        _x = Math.cos( angle ) * magnitude;
        _y = Math.sin( angle ) * magnitude;
    }
    
    public void setMagnitude( double magnitude ) {
        setMagnitudeAngle( magnitude, _angle );
    }
    
    public double getMagnitude() {
        return _magnitude;
    }
    
    public void setAngle( double angle ) {
        setMagnitudeAngle( _magnitude, angle );
    }
    
    public double getAngle() {
        return _angle;
    }
    
    //----------------------------------------------------------------------------
    // Operators
    //----------------------------------------------------------------------------
    
    public void add( Vector2D v ) {
        setXY( _x + v.getX(), _y + v.getY() );
    }
    
    public void subtract( Vector2D v ) {
        setXY( _x - v.getX(), _y - v.getY() );
    }
    
    public void scale( double scale ) {
        setXY( _x * scale, _y * scale );
    }
    
    public void rotate( double angle ) {
        setAngle( _angle + angle );
    }
    
    //----------------------------------------------------------------------------
    // Object overrides
    //----------------------------------------------------------------------------

    public boolean equals( Object object ) {
        boolean result = false;
        if ( getClass() == object.getClass() ) {
            Vector2D v = (Vector2D) object;
            result = ( _x == v.getX() && _y == v.getY() );
        }
        return result;
    }

    public String toString() {
        return getClass().getName() + "[" + _x + ", " + _y + "]";
    }
}
